### PR TITLE
load configuration

### PR DESCRIPTION
**Pull request by kevodwyer**
supports:
- loading configuration from an existing .ipfs install
- loading configuration from a ipfs-nucleus install
- creating a new configuration for nabu

### CODE REVIEW

File: src/main/java/org/peergos/Server.java [hunk 1]
```diff
@@ -13,106 +15,209 @@
...
-            throw new IllegalStateException("Unable to find ./ipfs/config file");
+            System.out.println("Unable to find ./ipfs/config file. Creating default config");
+            String contents = defaultConfig();
+            Files.write(configFilePath, contents.getBytes(), StandardOpenOption.CREATE);
+            return new Config((HashMap) JSONParser.parse(contents));
         }
         String contents = Files.readString(configFilePath);
         return new Config((HashMap) JSONParser.parse(contents));
     }
+    private String defaultConfig() {
```
Comment (ianopolous) [inline src/main/java/org/peergos/Server.java:130]: Cleaner to return the map here which can be easily added to I think, and just serialise it in time for writing, and avoid the serde loop.

Comment (kevodwyer) [reply src/main/java/org/peergos/Server.java]: done

File: src/main/java/org/peergos/Server.java [hunk 2]
```diff
@@ -13,106 +15,209 @@
...
-//            KademliaController bootstrap = builder.getWanDht().get().dial(node, bootstrapNode).getController().join();
-
-            APIService localAPI = new APIService();
-            String apiStr = config.getProperty("Addresses", "API");
-            String apiAddress = "/ip4/127.0.0.1/tcp/";
-            String apiPortStr = apiStr.startsWith(apiAddress) ? apiStr.substring(apiAddress.length()) : null;
-            if (apiPortStr == null) {
-                throw new IllegalStateException("Unable to determine API port");
+    public Server() throws Exception {
+        String home = System.getenv("HOME");
```
Comment (ianopolous) [inline src/main/java/org/peergos/Server.java:36]: Can we support he env var that ipfs uses to set the base .ipfs dir?

File: src/main/java/org/peergos/Server.java [hunk 3]
```diff
@@ -13,106 +15,209 @@
...
-            });
-            Runtime.getRuntime().addShutdownHook(shutdownHook);
-        } catch (Exception e) {
-            throw new RuntimeException(e);
+        } else if(blocksDirectory.isFile()) {
+            throw new IllegalStateException("Unable to create blocks directory");
         }
+        FileBlockstore blocks = new FileBlockstore(blocksPath);
+
+        int hostPort = 6001; // 6001
```
Comment (ianopolous) [inline src/main/java/org/peergos/Server.java:70]: we should get this from the config

File: src/main/java/org/peergos/Server.java (hunk 2 above)
Comment (kevodwyer) [reply src/main/java/org/peergos/Server.java]: done

File: src/main/java/org/peergos/Server.java (hunk 3 above)
Comment (kevodwyer) [reply src/main/java/org/peergos/Server.java:68]: i was wondering which line in the config it referes to. Can you point me to it? No occurence of 6001 afaict

Comment (kevodwyer) [reply src/main/java/org/peergos/Server.java]: i extract the Addresses -> API MultiAddress for use by the java-ipfs-http-client RPC client.

File: src/main/java/org/peergos/Server.java [hunk 4]
```diff
@@ -13,106 +15,222 @@
...
+    private Config readConfig(Path configPath) throws IOException {
+        Path configFilePath = configPath.resolve("config");
         File configFile = configFilePath.toFile();
         if (! configFile.exists()) {
-            throw new IllegalStateException("Unable to find ./ipfs/config file");
+            System.out.println("Unable to find ./ipfs/config file. Creating default config");
+            Config config = new Config(defaultConfig());
+            String contents = config.prettyPrint();
+            Files.write(configFilePath, contents.getBytes(), StandardOpenOption.CREATE);
+            return new Config((Map) JSONParser.parse(contents));
```
Comment (ianopolous) [inline src/main/java/org/peergos/Server.java:139]: can just return config

Comment (kevodwyer) [reply src/main/java/org/peergos/Server.java]: done

**Review by ianopolous — APPROVED** [review]

**Pull request merged**